pyphare/pyphare/pharesee/hierarchy.py

```diff
@@ -243,67 +268,163 @@ def overlap_mask(x, level, qty):
 
             pdata = patch.patch_datas[qty]
-            ghosts_nbr = pdata.ghosts_nbr[0]
-            patch_values = pdata.dataset[ghosts_nbr:-ghosts_nbr]
-            x = pdata.x[ghosts_nbr:-ghosts_nbr]
+            ghosts_nbr = pdata.ghosts_nbr
+
+            data = pdata.dataset[ghosts_nbr[0]:-ghosts_nbr[0]]
+            x = pdata.x[ghosts_nbr[0]:-ghosts_nbr[0]]
 
             if ilvl == hierarchy.finest_level(time):
                 if ip == 0:
-                    final = patch_values
+                    final_data = data
                     final_x = x
-                    last_x = x
                 else:
-                    if are_adjacent(last_x, x):
-                        first_node = 1
-                    else:
-                        first_node = 0
-
-                    final = np.concatenate((final, patch_values[first_node:]))
-                    final_x = np.concatenate((final_x, x[first_node:]))
-                    last_x = x[first_node:]
+                    final_data = np.concatenate((final_data, data))
+                    final_x = np.concatenate((final_x, x))
 
             else:
-                is_overlaped = overlap_mask(x, hierarchy.level(ilvl+1, time), qty)
-                unique_values = patch_values[~is_overlaped]
-                unique_x = x[~is_overlaped]
+                is_overlaped = overlap_mask_1d(x, pdata.dl,\
+                               hierarchy.level(ilvl+1, time), qty)
 
-                if unique_x.size:
-                    if are_adjacent(last_x, unique_x):
-                        first_node = 1
-                    else:
-                        first_node = 0
+                finest_data = data[~is_overlaped]
+                finest_x = x[~is_overlaped]
+
+                final_data = np.concatenate((final_data, finest_data))
+                final_x = np.concatenate((final_x, finest_x))
+
+    return final_data, final_x
+
+
+
+
+def flat_finest_field_2d(hierarchy, qty, time=None):
+
+    lvl = hierarchy.levels(time)
+
+    for ilvl in range(hierarchy.finest_level(time)+1)[::-1]:
+        patches = lvl[ilvl].patches
 
-                    last_x = unique_x[first_node:]
+        for ip, patch in enumerate(patches):
 
-                    tmp_x = np.concatenate((final_x, unique_x[first_node:]))
-                    final_x = np.sort(tmp_x)
+            pdata = patch.patch_datas[qty]
+            ghosts_nbr = pdata.ghosts_nbr
 
-                    # ok now we need to insert unique_values in
-                    # final. unique_values should be inserted in final
-                    # in the same order unique_x should be in final_x
-                    # i.e. with the indexing that sorts tmp_x
+            dset = np.array(pdata.dataset).reshape((pdata.x.shape[0],\
+                                                    pdata.y.shape[0]))
+            data = dset[ghosts_nbr[0]:-ghosts_nbr[0],\
+                        ghosts_nbr[1]:-ghosts_nbr[1]]
```

Comment:
I think this reshape shouldn't be necessary, do you need the copy?

otherwise I think the following suffices.

```
            data = pdata.dataset[ghosts_nbr[0]:-ghosts_nbr[0],\
                        ghosts_nbr[1]:-ghosts_nbr[1]]
```

eventually once the py_2d PR is merged this could also be replaced with 

```
            data = pdata[patch.box]
```

Reply:
I tried the solution you proposed unsuccessfully. The quantity `pdata.dataset` is is a flattened dataset, and does not accept a 2 indices (`ValueError: 2 indexing arguments for 1 dimensions`). I did not found a reshape method for such type of object. I agree that the copy is unneeded, but I do not how to do without.

Reply:
The array should not be flat, there was a PR to put diagnostic arrays in the correct shape from C++, if it is flat that's something to address I think

Reply:
You're right! the first expression is now ok without reshape.
but the second one doesn't work, even in 1d... I did not go through this Py_2d PR, so it's not clear to me if it should or not.

Reply:
not yet, and I think it should not deprecate the first notation anyway so I'd say you're good.